repair/repair.cc

```diff
@@ -1920,20 +1922,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            std::unordered_set<gms::inet_address> dc_nodes;
+            sstring ks_source_dc;
+            if (ignored_in_my_dc < 0) {
+                size_t ignored_in_my_dc = 0;
+                topology.for_each_node([&] (const locator::node* node) {
+                    if (node->dc_rack().dc != source_dc) {
+                        return;
+                    }
+                    if (ignore_nodes.contains(node->endpoint())) {
+                        ++ignored_in_my_dc;
+                    } else {
+                        dc_nodes.insert(node->endpoint());
+                    }
+                });
+            }
+            // Allow repairing in the dc only if there are enough replicas
+            if (!source_dc.empty() && !dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    if (ignored_in_my_dc) {
+                        // Don't take any chances if there are ignored nodes in this dc
+                        break;
+                    }
+                    auto opts = strat.get_config_options();
+                    auto rf = opts.contains(source_dc) ? std::atoi(opts[source_dc].c_str()) : 0;
+                    if (rf > 1) {
```

Comment:
I think if the source DC has RF =0, we should reject the invalid request and report to user.  Instead of silently falling back to use all DCs.

Reply:
What can the user do with such an error?
The source_dc is selected automatically by scylla.

Reply:
when a user runs rebuild, it can provides a source dc from the cmdline.

Reply:
For rebuild we probably should fail. For replace, we must choose the safe behavior.

Reply:
It's easy to reject when rf=0, but what if the user made a mistake and used source_dc when there aren't enough replicas to ensure quorum?
We can just respect that and print a warning about the risk.
if we had a --force option as well we could fail in this case too, unless --force was given.